full-service/src/service/sync.rs

```diff
@@ -446,28 +365,30 @@ pub fn decode_subaddress_index(
 /// should only be done on tx-outs that have already had their amounts decoded.
 /// If this fails, then the transaction is "orphaned", meaning we haven't
 /// generated the correct subaddress yet.
+///
+/// Key images will only be generated if the `account_key` is provided.
 pub fn decode_subaddress_and_key_image(
     tx_out: &TxOut,
-    account_key: &AccountKey,
+    view_private_key: &RistrettoPrivate,
+    account_key: Option<&AccountKey>,
     conn: Conn,
 ) -> (Option<u64>, Option<KeyImage>) {
-    let tx_public_key = match RistrettoPublic::try_from(&tx_out.public_key) {
-        Ok(k) => k,
-        Err(_) => return (None, None),
-    };
-
-    let subaddress_index = decode_subaddress_index(tx_out, account_key.view_private_key(), conn);
-
-    let key_image = if let Some(subaddress_i) = subaddress_index {
-        let onetime_private_key = recover_onetime_private_key(
-            &tx_public_key,
-            account_key.view_private_key(),
-            &account_key.subaddress_spend_private(subaddress_i),
-        );
-        Some(KeyImage::from(&onetime_private_key))
-    } else {
-        None
-    };
+    let subaddress_index = decode_subaddress_index(tx_out, view_private_key, conn);
+
+    let key_image = account_key.and_then(|account_key| {
```

Comment:
TBH i'm not fond of this logic and extension to this function, but it seemed like the minimal change to reduce the excessive code duplication in `sync_account_next_chunk()(`